drv/psc-psu-update/src/main.rs

```diff
@@ -0,0 +1,389 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! Server for updating all PSUs to the contained binary payload.
+//!
+//! We have the capacity to dynamically update the MWOCP68 power supply units
+//! connected to the PSC.  This update does not involve any interruption of
+//! the PSU while it is being performed, but necessitates a reset of the PSU,
+//! which will result in a drop of its own output.  We want these updates to
+//! be automatic and autonomous; there is little that the control plane can
+//! know that we do not know -- and even less for the operator.
+//!
+//! This task contains within it a payload that is the desired firmware image
+//! (`MWOCP68_FIRMWARE_PAYLOAD`), along with the `MFR_REVISION` that that
+//! pyaload represents (``MWOCP68_FIRMWARE_VERSION`).  This task will check
+//! every PSU periodically to see if the PSU's firmware revision matches the
+//! payload revision; if it doesn't (or rather, until it does), an attempt
+//! will be made to update the PSU.  Each PSU will be updated sequentially:
+//! while we can expect a properly configured and operating rack to support
+//! the loss of any one PSU, we do not want to induce the loss of more than
+//! one simultaneously due to update.  If an update fails, the update of that
+//! PSU will be exponentially backed off and repeated (up to a backoff of
+//! about once per day).  Note that we will continue to check PSUs that we
+//! have already updated should they be replaced with a PSU with downrev
+//! firmware.  The state of this task can be ascertained by looking at the
+//! `PSU` variable (which contains all of the per-PSU state) as well as the
+//! ring buffer.
+//!
+
+#![no_std]
+#![no_main]
+
+use drv_i2c_api::*;
+use drv_i2c_devices::mwocp68::{
+    Error as Mwocp68Error, FirmwareRev, Mwocp68, SerialNumber, UpdateState,
+};
+use ringbuf::*;
+use static_cell::ClaimOnceCell;
+use userlib::*;
+
+use core::ops::Add;
+
+task_slot!(I2C, i2c_driver);
+
+const TIMER_INTERVAL_MS: u64 = 10_000;
+
+use i2c_config::devices;
+
+#[cfg(any(target_board = "psc-b", target_board = "psc-c"))]
+static DEVICES: [fn(TaskId) -> I2cDevice; 6] = [
+    devices::mwocp68_psu0mcu,
+    devices::mwocp68_psu1mcu,
+    devices::mwocp68_psu2mcu,
+    devices::mwocp68_psu3mcu,
+    devices::mwocp68_psu4mcu,
+    devices::mwocp68_psu5mcu,
+];
+
+static PSU: ClaimOnceCell<[Psu; 6]> = ClaimOnceCell::new(
+    [Psu {
+        last_checked: None,
+        present: None,
+        power_good: None,
+        serial_number: None,
+        firmware_matches: None,
+        firmware_revision: None,
+        update_started: None,
+        update_succeeded: None,
+        update_failure: None,
+        update_backoff: None,
+    }; 6],
+);
+
+#[derive(Copy, Clone, Debug, PartialEq, counters::Count)]
+enum Trace {
+    None,
```

Comment:
should probably skip this.
```suggestion
    #[count(skip)]
    None,
```